Isolate loading and deployment of plugins

I'm trying to bundle builtin VS Code extensions: https://github.com/theia-ide/theia/issues/3956#issuecomment-451946193

Some of the extensions are mispackaged, or there are bugs in our code to parse manifest of certain extensions. Whenever the plugin system stumbles at such extension it throws and interrupts processing of consequent extensions.

Some call stacks which I get:
```
root ERROR Uncaught Exception:  SyntaxError: Unexpected token P in JSON at position 28873
root ERROR SyntaxError: Unexpected token P in JSON at position 28873
    at JSON.parse (<anonymous>)
    at HostedPluginReader.getPluginMetadata (/workspace/vscode-builtin-extensions/node_modules/@theia/plugin-ext/lib/hosted/node/plugin-reader.js:71:27)
    at /workspace/vscode-builtin-extensions/node_modules/@theia/plugin-ext/lib/hosted/node/plugin-service.js:119:47
    at Array.forEach (<anonymous>)
    at HostedPluginServerImpl.deployBackendPlugins (/workspace/vscode-builtin-extensions/node_modules/@theia/plugin-ext/lib/hosted/node/plugin-service.js:118:24)
    at PluginDeployerImpl.<anonymous> (/workspace/vscode-builtin-extensions/node_modules/@theia/plugin-ext/lib/main/node/plugin-deployer-impl.js:192:41)
    at step (/workspace/vscode-builtin-extensions/node_modules/@theia/plugin-ext/lib/main/node/plugin-deployer-impl.js:56:23)
    at Object.next (/workspace/vscode-builtin-extensions/node_modules/@theia/plugin-ext/lib/main/node/plugin-deployer-impl.js:37:53)
    at /workspace/vscode-builtin-extensions/node_modules/@theia/plugin-ext/lib/main/node/plugin-deployer-impl.js:31:71
    at new Promise (<anonymous>)
```
```
root ERROR [hosted-plugin: 9591] { Error: Cannot find module '/tmp/vscode-unpacked/configuration-editing-1.0.0.vsix/extension/out/extension'
    at Function.Module._resolveFilename (module.js:548:15)
    at Function.Module._load (module.js:475:25)
    at Function.module._load (/workspace/vscode-builtin-extensions/node_modules/@theia/plugin-ext-vscode/lib/node/plugin-vscode-init.js:94:33)
    at Module.require (module.js:597:17)
    at require (internal/module.js:11:18)
    at Object.loadPlugin (/workspace/vscode-builtin-extensions/node_modules/@theia/plugin-ext/lib/hosted/node/plugin-host-rpc.js:68:28)
    at PluginManagerExtImpl.$init (/workspace/vscode-builtin-extensions/node_modules/@theia/plugin-ext/lib/plugin/plugin-manager.js:123:44)
    at RPCProtocolImpl.doInvokeHandler (/workspace/vscode-builtin-extensions/node_modules/@theia/plugin-ext/lib/api/rpc-protocol.js:173:23)
    at RPCProtocolImpl.invokeHandler (/workspace/vscode-builtin-extensions/node_modules/@theia/plugin-ext/lib/api/rpc-protocol.js:158:41)
    at RPCProtocolImpl.receiveRequest (/workspace/vscode-builtin-extensions/node_modules/@theia/plugin-ext/lib/api/rpc-protocol.js:122:45) code: 'MODULE_NOT_FOUND' }
```

I'm going to look into fixing packaging of VS Code extensions and individual parsing issues. Someone else can look into adding  `try - catch` at cases where we loop over plugins.

--- a/packages/plugin-ext/src/hosted/node/plugin-reader.ts
+++ b/packages/plugin-ext/src/hosted/node/plugin-reader.ts
@@ -13,14 +13,18 @@
  *
  * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
  ********************************************************************************/
-import { BackendApplicationContribution } from '@theia/core/lib/node/backend-application';
-import { inject, injectable, optional, multiInject } from 'inversify';
+
+ // tslint:disable:no-any
+
+import * as path from 'path';
+import * as fs from 'fs-extra';
 import * as express from 'express';
-import * as fs from 'fs';
-import { resolve } from 'path';
-import { MetadataScanner } from './metadata-scanner';
-import { PluginMetadata, PluginPackage, getPluginId, MetadataProcessor } from '../../common/plugin-protocol';
 import { ILogger } from '@theia/core';
+import { inject, injectable, optional, multiInject } from 'inversify';
+import { Deferred } from '@theia/core/lib/common/promise-util';
+import { BackendApplicationContribution } from '@theia/core/lib/node/backend-application';
+import { PluginMetadata, getPluginId, MetadataProcessor } from '../../common/plugin-protocol';
+import { MetadataScanner } from './metadata-scanner';
 
 @injectable()
 export class HostedPluginReader implements BackendApplicationContribution {
@@ -31,7 +35,7 @@ export class HostedPluginReader implements BackendApplicationContribution {
     @inject(MetadataScanner)
     private readonly scanner: MetadataScanner;
 
-    private plugin: PluginMetadata | undefined;
+    private readonly hostedPlugin = new Deferred<PluginMetadata | undefined>();
 
     @optional()
     @multiInject(MetadataProcessor) private readonly metadataProcessors: MetadataProcessor[];
@@ -42,15 +46,8 @@ export class HostedPluginReader implements BackendApplicationContribution {
     private pluginsIdsFiles: Map<string, string> = new Map();
 
     initialize(): void {
-        if (process.env.HOSTED_PLUGIN) {
-            let pluginPath = process.env.HOSTED_PLUGIN;
-            if (pluginPath) {
-                if (!pluginPath.endsWith('/')) {
-                    pluginPath += '/';
-                }
-                this.plugin = this.getPluginMetadata(pluginPath);
-            }
-        }
+        this.doGetPluginMetadata(process.env.HOSTED_PLUGIN)
+            .then(this.hostedPlugin.resolve.bind(this.hostedPlugin));
     }
 
     configure(app: express.Application): void {
@@ -68,29 +65,46 @@ export class HostedPluginReader implements BackendApplicationContribution {
         });
     }
 
-    getPluginMetadata(path: string): PluginMetadata | undefined {
-        if (!path.endsWith('/')) {
-            path += '/';
+    async getPluginMetadata(pluginPath: string): Promise<PluginMetadata | undefined> {
+        const plugin = await this.doGetPluginMetadata(pluginPath);
+        if (plugin) {
+            const hostedPlugin = await this.getPlugin();
+            if (hostedPlugin && hostedPlugin.model.name === plugin.model.name) {
+                // prefer hosted plugin
+                return undefined;
+            }
         }
-        const packageJsonPath = path + 'package.json';
-        if (!fs.existsSync(packageJsonPath)) {
+        return plugin;
+    }
+
+    /**
+     * MUST never throw to isolate plugin deployment
+     */
+    protected async doGetPluginMetadata(pluginPath: string | undefined) {
+        try {
+            if (!pluginPath) {
+                return undefined;
+            }
+            if (!pluginPath.endsWith('/')) {
+                pluginPath += '/';
+            }
+            return await this.loadPluginMetadata(pluginPath);
+        } catch (e) {
+            this.logger.error(`Failed to load plugin metadata from "${pluginPath}"`, e);
             return undefined;
         }
+    }
 
-        let rawData = fs.readFileSync(packageJsonPath).toString();
-        rawData = this.localize(rawData, path);
-
-        const plugin: PluginPackage = JSON.parse(rawData);
-        plugin.packagePath = path;
-        const pluginMetadata = this.scanner.getPluginMetadata(plugin);
-        if (this.plugin && this.plugin.model && this.plugin.model.name === pluginMetadata.model.name) {
-            // prefer hosted plugin
+    protected async loadPluginMetadata(pluginPath: string): Promise<PluginMetadata | undefined> {
+        const manifest = await this.loadManifest(pluginPath);
+        if (!manifest) {
             return undefined;
         }
+        manifest.packagePath = pluginPath;
+        const pluginMetadata = this.scanner.getPluginMetadata(manifest);
         if (pluginMetadata.model.entryPoint.backend) {
-            pluginMetadata.model.entryPoint.backend = resolve(path, pluginMetadata.model.entryPoint.backend);
+            pluginMetadata.model.entryPoint.backend = path.resolve(pluginPath, pluginMetadata.model.entryPoint.backend);
         }
-
         if (pluginMetadata) {
             // Add post processor
             if (this.metadataProcessors) {
@@ -98,29 +112,61 @@ export class HostedPluginReader implements BackendApplicationContribution {
                     metadataProcessor.process(pluginMetadata);
                 });
             }
-            this.pluginsIdsFiles.set(getPluginId(pluginMetadata.model), path);
+            this.pluginsIdsFiles.set(getPluginId(pluginMetadata.model), pluginPath);
         }
-
         return pluginMetadata;
     }
 
-    private localize(rawData: string, pluginPath: string): string {
-        const nlsPath = pluginPath + 'package.nls.json';
-        if (fs.existsSync(nlsPath)) {
-            const nlsMap: {
-                [key: string]: string
-            } = require(nlsPath);
-            for (const key of Object.keys(nlsMap)) {
-                const value = nlsMap[key].replace(/\"/g, '\\"');
-                rawData = rawData.split('%' + key + '%').join(value);
+    async getPlugin(): Promise<PluginMetadata | undefined> {
+        return this.hostedPlugin.promise;
+    }
+
+    protected async loadManifest(pluginPath: string): Promise<any> {
+        const [manifest, translations] = await Promise.all([
+            fs.readJson(path.join(pluginPath, 'package.json')),
+            this.loadTranslations(pluginPath)
+        ]);
+        return manifest && translations && Object.keys(translations).length ?
+            this.localize(manifest, translations) :
+            manifest;
+    }
+
+    protected async loadTranslations(pluginPath: string): Promise<any> {
+        try {
+            return await fs.readJson(path.join(pluginPath, 'package.nls.json'));
+        } catch (e) {
+            if (e.code !== 'ENOENT') {
+                throw e;
             }
+            return {};
         }
-
-        return rawData;
     }
 
-    getPlugin(): PluginMetadata | undefined {
-        return this.plugin;
+    protected localize(value: any, translations: {
+        [key: string]: string
+    }): any {
+        if (typeof value === 'string') {
+            const match = HostedPluginReader.NLS_REGEX.exec(value);
+            return match && translations[match[1]] || value;
+        }
+        if (Array.isArray(value)) {
+            const result = [];
+            for (const item of value) {
+                result.push(this.localize(item, translations));
+            }
+            return result;
+        }
+        if (typeof value === 'object') {
+            const result: { [key: string]: any } = {};
+            // tslint:disable-next-line:forin
+            for (const propertyName in value) {
+                result[propertyName] = this.localize(value[propertyName], translations);
+            }
+            return result;
+        }
+        return value;
     }
+
+    static NLS_REGEX = /^%([\w\d.-]+)%$/i;
 
 }

--- a/packages/plugin-ext/src/hosted/node/plugin-service.ts
+++ b/packages/plugin-ext/src/hosted/node/plugin-service.ts
@@ -26,7 +26,6 @@ import { ExtPluginApiProvider, ExtPluginApi } from '../../common/plugin-ext-api-
 
 @injectable()
 export class HostedPluginServerImpl implements HostedPluginServer {
-
     @inject(ILogger)
     protected readonly logger: ILogger;
     @inject(HostedPluginsManager)
@@ -58,8 +57,8 @@ export class HostedPluginServerImpl implements HostedPluginServer {
     setClient(client: HostedPluginClient): void {
         this.hostedPlugin.setClient(client);
     }
-    getHostedPlugin(): Promise<PluginMetadata | undefined> {
-        const pluginMetadata = this.reader.getPlugin();
+    async getHostedPlugin(): Promise<PluginMetadata | undefined> {
+        const pluginMetadata = await this.reader.getPlugin();
         if (pluginMetadata) {
             this.hostedPlugin.runPlugin(pluginMetadata.model);
         }
@@ -78,16 +77,14 @@ export class HostedPluginServerImpl implements HostedPluginServer {
     }
 
     // need to run a new node instance with plugin-host for all plugins
-    deployFrontendPlugins(frontendPlugins: PluginDeployerEntry[]): Promise<void> {
-        // get metadata
-        frontendPlugins.forEach(frontendPluginDeployerEntry => {
-            const pluginMetadata = this.reader.getPluginMetadata(frontendPluginDeployerEntry.path());
-            if (pluginMetadata) {
-                this.currentFrontendPluginsMetadata.push(pluginMetadata);
-                this.logger.info('HostedPluginServerImpl/ asking to deploy the frontend Plugin', frontendPluginDeployerEntry.path(), 'and model is', pluginMetadata.model);
+    async deployFrontendPlugins(frontendPlugins: PluginDeployerEntry[]): Promise<void> {
+        for (const plugin of frontendPlugins) {
+            const metadata = await this.reader.getPluginMetadata(plugin.path());
+            if (metadata) {
+                this.currentFrontendPluginsMetadata.push(metadata);
+                this.logger.info(`Deploying frontend plugin "${metadata.model.name}@${metadata.model.version}" from "${metadata.model.entryPoint.frontend || plugin.path()}"`);
             }
-        });
-        return Promise.resolve();
+        }
     }
 
     getDeployedBackendMetadata(): Promise<PluginMetadata[]> {
@@ -95,20 +92,17 @@ export class HostedPluginServerImpl implements HostedPluginServer {
     }
 
     // need to run a new node instance with plugin-host for all plugins
-    deployBackendPlugins(backendPlugins: PluginDeployerEntry[]): Promise<void> {
+    async deployBackendPlugins(backendPlugins: PluginDeployerEntry[]): Promise<void> {
         if (backendPlugins.length > 0) {
             this.hostedPlugin.runPluginServer();
         }
-
-        // get metadata
-        backendPlugins.forEach(backendPluginDeployerEntry => {
-            const pluginMetadata = this.reader.getPluginMetadata(backendPluginDeployerEntry.path());
-            if (pluginMetadata) {
-                this.currentBackendPluginsMetadata.push(pluginMetadata);
-                this.logger.info('HostedPluginServerImpl/ asking to deploy the backend Plugin', backendPluginDeployerEntry.path(), 'and model is', pluginMetadata.model);
+        for (const plugin of backendPlugins) {
+            const metadata = await this.reader.getPluginMetadata(plugin.path());
+            if (metadata) {
+                this.currentBackendPluginsMetadata.push(metadata);
+                this.logger.info(`Deploying backend plugin "${metadata.model.name}@${metadata.model.version}" from "${metadata.model.entryPoint.backend || plugin.path()}"`);
             }
-        });
-        return Promise.resolve();
+        }
     }
 
     onMessage(message: string): Promise<void> {

--- a/packages/plugin-ext/src/main/node/plugin-deployer-impl.ts
+++ b/packages/plugin-ext/src/main/node/plugin-deployer-impl.ts
@@ -130,7 +130,7 @@ export class PluginDeployerImpl implements PluginDeployer {
     /**
      * deploy all plugins that have been accepted
      */
-    public async deployPlugins(): Promise<any> {
+    async deployPlugins(): Promise<any> {
         const acceptedPlugins = this.pluginDeployerEntries.filter(pluginDeployerEntry => pluginDeployerEntry.isAccepted());
         const acceptedFrontendPlugins = this.pluginDeployerEntries.filter(pluginDeployerEntry => pluginDeployerEntry.isAccepted(PluginDeployerEntryType.FRONTEND));
         const acceptedBackendPlugins = this.pluginDeployerEntries.filter(pluginDeployerEntry => pluginDeployerEntry.isAccepted(PluginDeployerEntryType.BACKEND));
@@ -147,11 +147,11 @@ export class PluginDeployerImpl implements PluginDeployer {
         const pluginPaths = acceptedBackendPlugins.map(pluginEntry => pluginEntry.path());
         this.logger.debug('local path to deploy on remote instance', pluginPaths);
 
-        // start the backend plugins
-        this.hostedPluginServer.deployBackendPlugins(acceptedBackendPlugins);
-        this.hostedPluginServer.deployFrontendPlugins(acceptedFrontendPlugins);
-        return Promise.resolve();
-
+        await Promise.all([
+            // start the backend plugins
+            this.hostedPluginServer.deployBackendPlugins(acceptedBackendPlugins),
+            this.hostedPluginServer.deployFrontendPlugins(acceptedFrontendPlugins)
+        ]);
     }
 
     /**

